### PR TITLE
feat(storage): staking support

### PR DESCRIPTION
**Pull request by nduchak**
This PR introduces listening and handling of Staking Contract Events

### CODE REVIEW

**Review by AuHau — APPROVED** [review]

File: src/cli/start.ts
Comment (AuHau) [inline src/cli/start.ts:117]: Why prepending with `application`? `application.app` does not look nice 😅

**Review by AuHau — CHANGES_REQUESTED** [review]
LGTM, except on the handler structure, that I don't like.

File: src/services/storage/handlers/stake.ts
Comment (AuHau) [inline src/services/storage/handlers/stake.ts:15]: Maybe a cleaner solution here would be [`findOrCreate`](https://sequelize.org/v5/class/lib/model.js~Model.html#static-method-findOrCreate)?

File: src/services/storage/handlers/stake.ts
Comment (AuHau) [inline src/services/storage/handlers/stake.ts:35]: Well, that does not really make sense, right? Can you get `Unstaked` event for `token` and `account` that did not call `Staked` before? I would keep the error here.

File: src/services/storage/index.ts
Comment (AuHau) [inline src/services/storage/index.ts:82]: No need for the `${}` part.
```suggestion
  const storageEventsEmitter = getEventsEmitterForService(STORAGE_MANAGER, eth, storageManagerContract.abi as AbiItem[])
  const stakingEventsEmitter = getEventsEmitterForService(STAKING, eth, stakingContract.abi as AbiItem[])
```

File: src/services/storage/index.ts
Comment (AuHau) [inline src/services/storage/index.ts:129]:
```suggestion
    if (!isServiceInitialized(STORAGE_MANAGER) || !isServiceInitialized(STAKING)) {
```

File: src/services/storage/index.ts
Comment (AuHau) [inline src/services/storage/index.ts:137]:
```suggestion
    const storageManagerEventsEmitter = getEventsEmitterForService(STORAGE_MANAGER, eth, storageManagerContract.abi as AbiItem[])
```

File: src/services/storage/index.ts
Comment (AuHau) [inline src/services/storage/index.ts:147]:
```suggestion
    const stakingEventsEmitter = getEventsEmitterForService(STAKING, eth, stakingContract.abi as AbiItem[])
```

File: src/services/storage/index.ts
Comment (AuHau) [inline src/services/storage/index.ts:50]: Great idea to different the loggers!

**Review by AuHau — CHANGES_REQUESTED** [review]
Few suggestions

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:15]: Good idea

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:35]: Yea, theoretically it can not. But let's keep it there if you don't mind

Comment (AuHau) [reply src/services/storage/handlers/stake.ts:35]: Hmmm, I don't like this. Because you are making invalid state a valid one, instead allowing us to detect it (as the error will be logged). Yes there is nothing you can do really about it (and in those cases the right thing to do is usually error-out), but we should find out about these problems and investigate.

Out of curiousity I have wrote some tests for this case in to the SC: https://github.com/rsksmart/rif-marketplace-storage/pull/127

And also with this approach you will be getting actually negative total amount...

So I would say, throw an error ;-)

**Comment by AuHau** [comment]
Btw. this PR is blocked until the Staking PR is merged and released :-)

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:35]: I am confused, as I am throwing an error :)

Comment (AuHau) [reply src/services/storage/handlers/stake.ts:35]: Oh, sorry, you are rightfully confused 😅 Not sure why I thought that you are actually creating new entry there 😳
Sorry! My bad 🙈

**Review by AuHau — COMMENTED** [review]
LGTM, now we only have to release the Staking part of the SC.

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:35]: Probably that's because of force push which made before, so it's my bad=)

**Comment by nduchak** [comment]
Blocked by: https://github.com/rsksmart/rif-marketplace-storage/pull/112

**Comment by nduchak** [comment]
Closes #286 284

File: src/services/storage/handlers/stake.ts
Comment (AuHau) [inline src/services/storage/handlers/stake.ts:16]: Hmmmm, shouldn't this be specified in config?

File: src/services/storage/handlers/stake.ts
Comment (AuHau) [inline src/services/storage/handlers/stake.ts:232]: Can we make this function async please?

Comment (AuHau) [reply src/services/storage/handlers/stake.ts:232]: And also `token` is not really token but `tokenContractAddress` here no?

File: src/services/storage/handlers/stake.ts
Comment (AuHau) [inline src/services/storage/handlers/stake.ts:238]: Ufff, so you are calling the ERC20 contract here to resolve for given token address its symbol here? I am not sure this is a good approach. How fast is this call? Also, how do you make sure that this is actually value defined in `SupportedTokens`? 

I would say first of all we want have defined list of tokens that we support (most probably already mentioned `config` should be good place to have it there) and since we gonna do that we can add their address as well? We already do this for every other interaction with blockchain, so I would say it is standard and it will saves us lot of this work, that I am not sure if actually bring many benefits.

File: test/services/storage/models.spec.ts
Comment (AuHau) [inline test/services/storage/models.spec.ts:213]:
```suggestion
    it('should aggregate total stakes for offers and order by stakes', async () => {
```

File: src/services/storage/models/stake.model.ts
Comment (AuHau) [inline src/services/storage/models/stake.model.ts:6]: This is definitely place to have it. Either `config` or `definitions.ts`, as already mentioned I prefer the config with the token's address.

File: src/services/storage/models/offer.model.ts
Comment (AuHau) [inline src/services/storage/models/offer.model.ts:58]: Can you add JSDoc explaining what this does?

File: test/utils.ts
Comment (AuHau) [inline test/utils.ts:82]: Good catch, although don't understand why you have the above `if` with `for`? Shouldn't the `{recursive: true}` handle this for you?

**Review by AuHau — CHANGES_REQUESTED** [review]
I don't really agree with the approach for fetching the token's symbol. I would just use config and have configured tokens and their addresses. Simplicity FTW here...

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:232]: Yes, will rename it

File: src/services/storage/handlers/stake.ts
Comment (nduchak) [reply src/services/storage/handlers/stake.ts:238]: Hmm... The call is fast enough (it's static call), but I agree that tokens in the config would be better

File: src/services/storage/models/offer.model.ts
Comment (nduchak) [reply src/services/storage/models/offer.model.ts:58]: Will do

File: test/utils.ts
Comment (nduchak) [reply test/utils.ts:82]: `{recursive: true}` can not remove folder if folder contains some files

**Review by AuHau — APPROVED** [review]
LGTM, thanks for implementing the suggestions!

**Pull request merged**
